src/EVM/Expr.hs

```diff
@@ -1057,3 +1058,52 @@ evalProp prop =
       | l == r = PBool True
       | otherwise = o
     go p = p
+
+
+data ConstState = ConstState
+  { values :: Map.Map (Expr EWord) W256
+  , canBeSat :: Bool
+  }
+  deriving (Show)
+
+-- | Folds constants
+constFoldProp :: [Prop] -> ConstState
+constFoldProp ps = execState (mapM constProp ps) (ConstState mempty True)
+  where
+    constProp :: Prop -> State ConstState Prop
+    constProp prop = go (evalProp prop)
+    go :: Prop -> State ConstState Prop
```

Comment:
Why do we need the return value here? Don't we throw it away with the call to `execState` later?

Reply:
Yeah, good point. This has now been fixed!

Reply:
we can still simplify further here and change the type of `go` to `Prop -> State ConstState ()`

Reply:
Ah yess! I finally understood what you meant, thanks! OK, it's done! I'm resolving now. Can you please check and then we can merge!